cmd/ifparse.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * cmd/ifparse.c
+ * Purpose: Redisplay /e/n/i in alternative formats.
+ *
+ * Copyright (c) 2020 Ariadne Conill <ariadne@dereferenced.org>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * This software is provided 'as is' and without any warranty, express or
+ * implied.  In no event shall the authors be liable for any damages arising
+ * from the use of this software.
+ */
+
+#define _GNU_SOURCE
+#include <fnmatch.h>
+#include <stdio.h>
+#include <string.h>
+#include <getopt.h>
+#include "libifupdown/libifupdown.h"
+
+#ifdef CONFIG_YAML
+# include "libifupdown/yaml-base.h"
+# include "libifupdown/yaml-writer.h"
+#endif
+
+#include "cmd/multicall.h"
+#include "cmd/pretty-print-iface.h"
+
+static bool show_all = false;
+static bool allow_undefined = false;
+
+static void
+set_show_all(const char *arg)
+{
+	(void) arg;
+
+	show_all = true;
+}
+
+static void
+set_allow_undefined(const char *arg)
+{
+	(void) arg;
+
+	allow_undefined = true;
+}
+
+#ifdef CONFIG_YAML
+static bool use_yaml = false;
+
+static void
+set_yaml(const char *arg)
+{
+	(void) arg;
+
+	use_yaml = true;
+}
+#endif
+
+static struct if_option local_options[] = {
+	{'A', "all", NULL, "show all interfaces", false, set_show_all},
+	{'U', "allow-undefined", NULL, "allow querying undefined (virtual) interfaces", false, set_allow_undefined},
+#ifdef CONFIG_YAML
+	{'Y', "yaml-raw", NULL, "reflect raw {iface, key, value} triples as YAML", false, set_yaml},
+#endif
+};
+
+static struct if_option_group local_option_group = {
+	.desc = "Program-specific options",
+	.group_size = ARRAY_SIZE(local_options),
+	.group = local_options
+};
+
+#ifdef CONFIG_YAML
+static void
+prettyprint_interface_yaml(struct lif_interface *iface)
+{
+	struct lif_yaml_node doc = {};
+
+	lif_yaml_document_init(&doc, "interfaces");
+
+	struct lif_yaml_node *iface_node = lif_yaml_node_new_list(iface->ifname);
+	lif_yaml_node_append_child(&doc, iface_node);
+
+	if (iface->is_auto)
+	{
+		struct lif_yaml_node *iface_entry_node = lif_yaml_node_new_boolean("auto", true);
+		lif_yaml_node_append_child(iface_node, iface_entry_node);
+	}
+
+	struct lif_node *iter;
+	LIF_DICT_FOREACH(iter, &iface->vars)
+	{
+		struct lif_dict_entry *entry = iter->data;
+		const char *value = entry->data;
+		char addr_buf[512];
+
+		if (!strcmp(entry->key, "address"))
+		{
+			struct lif_address *addr = entry->data;
+
+			if (!lif_address_unparse(addr, addr_buf, sizeof addr_buf, true))
+				continue;
```

Comment:
Do we need some kind of error handling / output here?

Reply:
What did you have in mind?

Reply:
I don't know really. I'm wondering what to do if `lif_address_unprase` fails. Looking at `lif_address_unparse` only `inet_ntop`can fail and that has only two documented error cases which we both control. So I guess "this is fine".